### PR TITLE
Cleanup user model

### PR DESCRIPTION
**Pull request by bgeuken**

### CODE REVIEW

File: src/api/app/models/user.rb [hunk 1]
```diff
@@ -53,14 +53,62 @@ class User < ActiveRecord::Base
...
+  validates_format_of :email,
+                      :with => %r{\A([\w\-\.\#\$%&!?*\'\+=(){}|~]+)@([0-9a-zA-Z\-\.\#\$%&!?*\'=(){}|~]+)+\z},
+                      :message => 'must be a valid email address.'
+
+  # We want to validate the format of the password and only allow alphanumeric
+  # and some punctiation characters.
+  # The format must only be checked if the password has been set and the record
+  # has not been stored yet and it has actually been set at all. Make sure you
+  # include this condition in your :if parameter to validates_format_of when
+  # overriding the password format validation.
```
Comment (hennevogel) [inline src/api/app/models/user.rb]: I don't understand the last two sentences

File: src/api/app/models/user.rb [hunk 2]
```diff
@@ -99,104 +147,70 @@ def self.default_user_state
...
+  # After validation, the password should be encrypted
+  after_validation(:on => :create) do
+    if errors.empty? and @new_password and !password.nil?
+      # generate a new 10-char long hash only Base64 encoded so things are compatible
+      self.password_salt = [Array.new(10){rand(256).chr}.join].pack('m')[0..9]
 
-  # After saving, we want to set the "@new_hash_type" value set to false
-  # again.
-  after_save '@new_hash_type = false'
+      # vvvvvv added this to maintain the password list for lighttpd
```
Comment (hennevogel) [inline src/api/app/models/user.rb]: What? lighttpd? huh?

Comment (bgeuken) [inline src/api/app/models/user.rb]: Duno. I just moved this up to the other callbacks. So the first 4 commits "Group foo" are just moving code, but not changing it.

@adrianschroeter Do you know if this is still valid / needed?
That line here ->
```
      # vvvvvv added this to maintain the password list for lighttpd
      write_attribute(:password_crypted, password.crypt('os'))
      #  ^^^^^^
```

File: src/api/app/models/user.rb (hunk 1 above)
Comment (bgeuken) [inline src/api/app/models/user.rb]: @hennevogel Would you mind fi I remove the last sentence at all? I wouldn't know when we ever overwrite the if parameter...
Same for the validates_length_of, line 103

File: src/api/app/models/user.rb (hunk 2 above)
Comment (adrianschroeter) [inline src/api/app/models/user.rb]: lighttpd is not really supported anymore.

File: src/api/app/models/user.rb [hunk 3]
```diff
@@ -261,27 +268,24 @@ def self.find_with_credentials(login, password)
       logger.debug( "No user found in database, creating" )
       logger.debug( "Email: #{ldap_info[0]}" )
       logger.debug( "Name : #{ldap_info[1]}" )
-      # Generate and store a fake pw in the OBS DB that no-one knows
-      chars = ["A".."Z", "a".."z", "0".."9"].collect { |r| r.to_a }.join
-      password = (1..24).collect { chars[rand(chars.size)] }.pack('a'*24)
+      # Generate and store a 24 char fake pw in the OBS DB that no-one knows
+      password = SecureRandom.base64
```
Comment (adrianschroeter) [inline src/api/app/models/user.rb]: Maybe a bit out-of-scope here, but the better approach would be not to set a random password, but some password which can not be used in real life. eg. we set it to "nil" and check for that on authentification in non-ldap case.
That would really eliminate any chance that someone can login if the LDAP config got disabled by accident.

File: src/api/app/models/user.rb (hunk 2 above)
Comment (bgeuken) [inline src/api/app/models/user.rb]: and we can remove the write_Attribute line...?

File: src/api/app/models/user.rb [hunk 4]
```diff
@@ -163,6 +163,207 @@ def self.password_hash_types
...
+  def self.find_by_email(email)
+    return where(:email => email).first
+  end
+
+  def self.realname_for_login(login)
+    User.find_by_login(login).realname
+  end
+
+  def self.fetch_field(person, field)
+    p = User.where(login: person).pluck(field)
```
Comment (adrianschroeter) [inline src/api/app/models/user.rb]: shouldn't we use here also find_by_login to avoid access of "deleted" accounts?

**Comment by adrianschroeter** [comment]
I would also like to drop password_confirmation and the two obsolete user states, but we can do that afterwards.

**Comment by adrianschroeter** [comment]
IMHO we can merge this.

**Pull request merged**